### ISSUE TITLE
auto detect device

### ISSUE THREAD

**Issue by garudaonekh** [madl3x/miflora_rbs#2]
Hi, can it auto detect new sensor? what is the registration process of new sensor? Can we move sensor from a station and register to new station?

**Comment by amircescu-gpfw** [comment]
> Hi, can it auto detect new sensor?

Yes, and it will be automatically added to the list of your MiFlora Devices. The station will scan periodically for new data, and new devices as soon as they are scanned. Once a new MiFlora device is found it will be added to the list with "Unknown" as name.

You can use this feature to "scan" for new MiFlora devices and write down their MAC address. After this step, you can construct your `devices.cfg` configuration (see below), and assign a Name, ID, and recommended limits for this plant sensor.

Make sure this option is set to `true` in `default_config.h` before compiling:
```
#define FLORA_DISCOVER_DEVICES                   true // automatically add devices that are not configured via devices.cfg
```

> what is the registration process of new sensor? 

Edit the `data/devices.cfg`file and add your MiFlora sensors. Then upload this file to the SPIFFS data partion of the ESP32. You can do this over the air, or via USB serial.

You can use the  `devices.cfg.sample` template file to construct you `devices.cfg` configuration.

One device in `devices.cfg` should look like this:
```
[C4:7C:8D:6A:5C:FF] ; MiFlora MAC address
name = Palmier ; MiFlora Name
id   = 1 ; Miflora ID (must be unique)
min_moisture=15
max_moisture=60
min_temperature=10
max_temperature=35
min_conductivity=350
max_conductivity=2000
min_illuminance=2000
max_illuminance=35000
min_rssi=-80
```
> Can we move sensor from a station and register to new station?

All stations should share the same `devices.cfg` configuration. Once they do, they will interoperate to collect and display data from these MiFlora devices, giving you the option to add more stations to increase the range where you can place your flowers.

Therefore, it doesn't matter what station is reading what sensor, as long as these stations are connected to the same MQTT server. 

For example, if Station A is reading values from Plant #1, and Station B is reading Plant #2, but Station A is not in range for reading Plant #2, and also Station B is not in range for reading Plant #1, we can still see both plants displayed on both stations, with the same values. Once Station A reads the moisture from Plant #1, it will share this value via MQTT with Station B, and viceversa.

This is the main idea behind this project.
Hope it helps!

**Comment by garudaonekh** [comment]
Thanks, how to do it via OTA? Can we configure this via MQTT? This mechanism is hard for normal user to use.

**Comment by garudaonekh** [comment]
I think the option to add new device on site is better. For example, click a button to add new device

**Comment by amircescu-gpfw** [comment]
> Thanks, how to do it via OTA? Can we configure this via MQTT? This mechanism is hard for normal user to use.

You can do this with either Arduino IDE or Visual Studio Code & Platform IO, the same that you will probably use for flashing and compiling the firmware.

I prefer the latter, so here's how you do it with Platform IO:
https://randomnerdtutorials.com/esp32-vs-code-platformio-spiffs/

> I think the option to add new device on site is better. For example, click a button to add new device

Clicking a button requires an HTTP server to run on the ESP, which adds more complexity.

Given the fact that you need to flash the firmware through the same process, so you need the setup anyway and is needed only once to configure all the MiFlora devices that you have, I don't find it that inconvenient yet.

**Comment by madl3x** [comment]
> Hi, can it auto detect new sensor?

The firmware will automatically detect new devices, if you have `FLORA_DISCOVER_DEVICES` set to `true` in `default_config.h` (currently the default value).

> Thanks, how to do it via OTA? 

Check out this [tutorial](https://randomnerdtutorials.com/esp32-vs-code-platformio-spiffs/) for how to update the "data" partition of an ESP32 using OTA and SPIFFS.

There are two templates in the `/data` folder: `config.cfg.sample` and `devices.cfg.sample`.
Rename these to `config.cfg` and `devices.cfg` and edit them accordingly.

`/data/config.cfg` is used so to overwrite the default settings set at compile time by `/include/default_config.h`.
This config file is useful if you want to use the same firmware image for multiple stations, and configure the stations differently, without recompiling. 

When the station boots, it looks for this `config.cfg` file to overwrite the default settings configured via `/include/default.config`. If it's not found, it will continue using the default settings from `default_config.h`.

`/data/devices.cfg` contains the list of MiFlora devices "known" by the station.
The configuration assigns a name, a unique ID, and the recommended limits for moisture, temperature, conductivity, illuminance, and RSSI, to each plant. 

> Can we configure this via MQTT? This mechanism is hard for normal user to use.

No. 

It is the same mechanism that you use for flashing the firmware, it only adds a marginal effort to write the "data" partition.
